File: solutions/2785. Sort Vowels in a String/2785_2.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
  def sortVowels(self, s: str) -> str:
    kVowels = 'AEIOUaeiou'
    count = [0] * 128
    for c in s:
      if c in kVowels:
        count[ord(c)] += 1

    ans = ""
    j = 0
    for i, c in enumerate(s):
      if c not in kVowels:
        ans += s[i]
      else:
        while count[ord(kVowels[j])] == 0:
          j += 1

        ans += kVowels[j]
        count[ord(kVowels[j])] -= 1

    return ans

File: solutions/2785. Sort Vowels in a String/2785_2.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
 public:
  string sortVowels(const string &s) {
    vector<int> count(128, 0);
    for (const char c : s)
      if (isVowel(c))
        ++count[c];

    string ans;
    int j = 0;
    for (int i = 0; i < s.size(); ++i)
      if (!isVowel(s[i])) {
        ans += s[i];
      } else {
        while (count[kVowels[j]] == 0)
          ++j;

        ans += kVowels[j];
        count[kVowels[j]]--;
      }

    return ans;
  }
 private:
  static constexpr string_view kVowels = "AEIOUaeiou";

  bool isVowel(const char& c) {
    return kVowels.find(c) != string_view::npos;
  }
};